### PR TITLE
Query Builder ADR

### PR DESCRIPTION
**Pull request by mikena-truss**
## Description

In the [Admin API Design](https://docs.google.com/document/d/1HAII3cFKAgVoiDKtkCB5RYi8A1FiGxLuOGJdAOoR_LI/edit#heading=h.87ohyyfofbcv) we defined a black box of a query builder. The reason being, admin queries need to be more flexible, but more consistent between models. Opposed to the current query patterns, where queries are written one off and don't follow a pattern across handlers.

This ADR discusses some possible options for building this query filter functionality.

### CODE REVIEW

File: docs/adr/0035-use-query-builder.md [hunk 1]
```diff
@@ -0,0 +1,267 @@
...
+  "col3": true,
+  "col4": false,
+  "col5": nil,
+  "col6": []string{"a", "b", "c"},
+}).ToSql()
+```
+
+* **Write a generic query interface that accepts any model**
+Another option is to write a query builder as a dependency to handlers/services.
+The proof of oncept is as follows:
```
Comment (lynzt) [inline docs/adr/0035-use-query-builder.md]: spelling: concept

File: docs/adr/0035-use-query-builder.md [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+  }
+  return query.All(model)
+}
+```
+
+And a service would utilize it as follows:
+
+```go
+type shipmentListQueryBuilder interface {
+  FetchMany(model interface{}, fitlers []services.QueryFilter) error
```
Comment (lynzt) [inline docs/adr/0035-use-query-builder.md]: spelling: filters

File: docs/adr/0035-use-query-builder.md [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+
+The handler (or middleware) marshalls those filters into a `QueryFilter`
+and passes them to the service
+
+## Decision Outcome
+
+* Chosen Alternative: **write a generic query builder**
+* The main motivators of this decision are:
+  * It accomplishes our desired feature set
+  * It is relatively simple to implement
```
Comment (lynzt) [inline docs/adr/0035-use-query-builder.md]: (thanks to you!! 🙇‍♀)

File: docs/adr/0035-use-query-builder.md [hunk 4]
```diff
@@ -0,0 +1,267 @@
...
+  invalidFields := make([]string, 0)
+  for _, f := range filters {
+    column, ok := getDBColumn(t, f.Column())
+    if !ok {
+      invalidFields = append(
+        invalidFields,
+        fmt.Sprintf("%s %s", f.Column(), f.Comparator()),
+      )
+    }
+    columnQuery := fmt.Sprintf("%s %s ?", column, comparator)
```
Comment (lynzt) [inline docs/adr/0035-use-query-builder.md]: I know this is pseudo code, but would these 2 lines be in an `else` block so they aren't run if we get an error returned from `getDBColumn()`?

Comment (mikena-truss) [reply docs/adr/0035-use-query-builder.md]: Good catch. I think a `continue might make sense`. I'll have to update that in the code too

**Pull request merged**